File: src/main/java/dev/adamko/gildedrose/java/item/ItemState.java
package dev.adamko.gildedrose.java.item;

import com.gildedrose.java.Item;
import org.jetbrains.annotations.NotNull;


/**
 * A Java implementation of {@link dev.adamko.gildedrose.item.ItemState}
 *
 * @see dev.adamko.gildedrose.item.ItemState
 */
public record ItemState(
    @NotNull
    String name,
    int sellIn,
    int quality,
    boolean isExpired
) {

  public ItemState(Item item) {
    this(item.name, item.sellIn, item.quality, isItemExpired(item));
  }

  public ItemState withQuality(int quality) {
    return new ItemState(this.name, this.sellIn, quality, this.isExpired);
  }

  public ItemState withSellIn(int sellIn) {
    return new ItemState(this.name, sellIn, this.quality, this.isExpired);
  }

  private static boolean isItemExpired(Item item) {
    return item.sellIn <= 0;
  }

}

File: src/main/kotlin/dev/adamko/gildedrose/item/ItemState.kt
package dev.adamko.gildedrose.item

import com.gildedrose.Item
import com.gildedrose.isExpired

/**
 * Freeze the state of an [Item], so it can be used safely in other contexts without mutating
 * the original.
 */
data class ItemState(
    val name: String,
    val sellIn: Int,
    val quality: Int,
    val isExpired: Boolean,
) {

  constructor(item: Item) : this(item.name, item.sellIn, item.quality, item.isExpired())

  companion object {
    fun Item.asState(): ItemState = ItemState(this)
  }
}
